frontend/src/metabase/lib/redux/typed-utils.ts

```diff
@@ -11,13 +12,26 @@ interface ThunkConfig {
 export const createAsyncThunk =
   createAsyncThunkOriginal.withTypes<ThunkConfig>();
 
+type AwaitedIfPromise<T> = T extends Promise<infer R> ? R : T;
+
 // similar to createAction but accepts a (redux-thunk style) thunk and dispatches based on whether
 // the promise returned from the thunk resolves or rejects, similar to redux-promise
-export function createThunkAction<TArgs extends any[]>(
-  actionType: string,
+export function createThunkAction<
+  TArgs extends any[],
+  TResult,
+  TActionType extends string,
+>(
+  actionType: TActionType,
   thunkCreator: (
     ...args: TArgs
-  ) => (dispatch: Dispatch, getState: GetState) => any,
-): (...args: TArgs) => any {
+  ) => (dispatch: ThunkDispatch<any, any, any>, getState: GetState) => TResult,
+): (
+  ...args: TArgs
+) => (
+  dispatch: ThunkDispatch<any, any, any>,
+  getState: GetState,
+) => Promise<{ type: TActionType; payload: AwaitedIfPromise<TResult> }> {
+  // @ts-expect-error - withAction is too hard to type correctly as it can accept both the payload or a thunk creator
```

Comment:
I initially tried to type `withAction`, but it was too much work and it would probably still have required some complex typing here too, so I just put a @ts-expect-error on it as the types that gets inferred by js are not correct